### ISSUE TITLE
Campaigns cannot be applied as a normal user when the code host is configured to use SSH to clone repositories

### ISSUE THREAD

**Issue by LawnGnome** [sourcegraph/sourcegraph#16888]
This occurs because `gitserver` will attempt to inject the user credential into the `ssh://` URL, which causes Git to be unable to parse the URL, after which a spurious "host not found" error is returned. This is a lie, but the issue is real.

#### Plan of attack

At a high level, we can deal with this by detecting code hosts with this configuration, then prompting users to add a public key (which we can generate and store) to their account when adding a PAT.

In terms of issues, this breaks down into:

* [Add an SSH key credential type.](https://github.com/sourcegraph/sourcegraph/issues/16890) (By making it a credential type, we can reuse existing progress towards securing secrets.)
* [Generate a private key on demand and store it in the database.](https://github.com/sourcegraph/sourcegraph/issues/16891)
* [Design the modified user settings flow to support this.](https://github.com/sourcegraph/sourcegraph/issues/16892)
* [Update the GraphQL schema to handle multiple credentials and detect when SSH keys are required.](https://github.com/sourcegraph/sourcegraph/issues/16893)
* [Update the user settings UI to provide the appropriate CTA and prompt with the public key.](https://github.com/sourcegraph/sourcegraph/issues/16894)
* [Update `gitserver` to use an SSH key credential instead of a PAT credential when pushing over SSH.](https://github.com/sourcegraph/sourcegraph/issues/16895)
* [Document how users can configure this.](https://github.com/sourcegraph/sourcegraph/issues/16896)

A bonus quest later _might_ be to automatically provision the key into the code host if the user scopes their PAT to allow this.

**Comment by mrnugget** [comment]
This seems like a lot of work and a lot of added complexity, so I have to ask: from what I could gather in Slack, the customer that ran into this _does allow pushing via HTTPs_, it's just that Sourcegraph thinks it should use SSH. Could a simpler solution be to let the customer configure that HTTPS should be used? (Off the top of my head: a `pushProtocol: string` option in the code host configuration, where the user can specify https/ssh and if it's ssh we say "not supported" for now?)

**Comment by mrnugget** [comment]
I dug into the documentation for using SSH keys a bit and found [this document here](https://docs.sourcegraph.com/admin/repo/auth#repositories-that-need-http-s-or-ssh-authentication).

Short version: SSH keys are kept on disk, or mounted as a Docker volume, or defined as Kubernetes secrets, or by mounting a `.ssh` folder into the `gitserver` container. This would introduce another way of adding and managing SSH keys, which again makes me think: do we really need to do this? I'm especially concerned about adding private keys to the database and then transferring them over the wire to gitserver to be used there. 

So, I think we should double-check with the customer whether they actually only allow push-via-SSH before we embark on this.

**Comment by eseliger** [comment]
During investigation last week we found out the following: A pure SSH-key credential might not be the key to unlock all the stuff here. For campaigns we always need a PAT in conjunction with the SSH-key. So the idea we came up with is extending the auth types by `BasicAuthWithSSHCredential` and `OAuthBearerTokenWithSSHCredential`, which would hold all required means of authentication for campaigns.
Then, we can extend the executor by detecting SSH, figuring out if the credential implements `SSHTokener` or something similar and use that for passing an SSH token to gitserver, which would then use `GIT_SSH_COMMAND=ssh -i tmppathtokey` to use the provided ssh token. If no credential is configured or the credential doesn't support ssh, it should bail out as before.

cc @mrnugget
